databroker/experimental/server_ext.py

```diff
@@ -0,0 +1,501 @@
+        )
+    data = await request.body()
+    entry.put_data(data)
+    return json_or_msgpack(request, None)
+
+
+@router.put("/dataframe/full/{path:path}")
+async def put_dataframe_full(
+    request: Request,
+    entry=Security(entry, scopes=["write:data", "write:metadata"]),
+):
+    if not isinstance(entry, WritingDataFrameAdapter):
+        raise HTTPException(
+            status_code=404, detail="This path cannot accept this dataframe."
+        )
+    data = await request.body()
+    entry.put_data(data)
+    return json_or_msgpack(request, None)
+
+
+def array_raise_if_inactive(method):
+    def inner(self, *args, **kwargs):
+        if self.array_adapter is None:
+            raise ValueError("Not active")
+        else:
+            return method(self, *args, **kwargs)
+
+    return inner
+
+
+def dataframe_raise_if_inactive(method):
+    def inner(self, *args, **kwargs):
+        if self.dataframe_adapter is None:
+            raise ValueError("Not active")
+        else:
+            return method(self, *args, **kwargs)
+
+    return inner
+
+
+class WritingArrayAdapter:
+    structure_family = "array"
+
+    def __init__(self, collection, directory, doc):
+        self.collection = collection
+        self.directory = directory
+        self.doc = Document(**doc)
+        self.array_adapter = None
+        if self.doc.data_url is not None:
+            path = self.doc.data_url.path
+            if platform == "win32" and path[0] == "/":
+                path = path[1:]
+
+            file = h5py.File(path)
+            dataset = file["data"]
+            self.array_adapter = ArrayAdapter(dask.array.from_array(dataset))
+        elif self.doc.data_blob is not None:
+            self.array_adapter = ArrayAdapter(dask.array.from_array(self.doc.data_blob))
+
+    @property
+    def structure(self):
+        return ArrayStructure.from_json(self.doc.structure)
+
+    @property
+    def metadata(self):
+        return self.doc.metadata
+
+    @array_raise_if_inactive
+    def read(self, *args, **kwargs):
+        return self.array_adapter.read(*args, **kwargs)
+
+    @array_raise_if_inactive
+    def read_block(self, *args, **kwargs):
+        return self.array_adapter.read_block(*args, **kwargs)
+
+    def microstructure(self):
+        return self.array_adapter.microstructure()
+
+    def macrostructure(self):
+        return self.array_adapter.macrostructure()
+
+    def put_data(self, body):
+        # Organize files into subdirectories with the first two
+        # charcters of the uid to avoid one giant directory.
+        path = self.directory / self.doc.uid[:2] / self.doc.uid
+        path.parent.mkdir(parents=True, exist_ok=True)
+        array = numpy.frombuffer(
+            body, dtype=self.doc.structure.micro.to_numpy_dtype()
+        ).reshape(self.doc.structure.macro.shape)
+        with h5py.File(path, "w") as file:
+            file.create_dataset("data", data=array)
+        self.collection.update_one(
```

Comment:
I think it's probably worth doing a check here that the update succeeded (one document matched, one document updated).